abb_hardware_interface/src/abb_system_position_only.cpp

```diff
@@ -12,243 +12,199 @@
-    RCLCPP_FATAL(
-      rclcpp::get_logger("ABBSystemPositionOnlyHardware"),
-      "Robot does not match expected number of joints");
-    return CallbackReturn::ERROR;
-  }
-
-  output_.mutable_robot()->mutable_joints()->mutable_position()->CopyFrom(current_positions_);
-  output_pos_.mutable_robot()->mutable_joints()->mutable_position()->CopyFrom(current_positions_);
-
-  for (size_t i = 0; i < hw_states_.size(); ++i) {
-    hw_states_[i] = current_positions_.values(i) / DEG_TO_RAD;
-    hw_commands_[i] = current_positions_.values(i) / DEG_TO_RAD;
-  }
-  RCLCPP_INFO(rclcpp::get_logger("ABBSystemPositionOnlyHardware"), "System Sucessfully started!");
-
-  return CallbackReturn::SUCCESS;
-}
 
-CallbackReturn ABBSystemPositionOnlyHardware::on_deactivate(
-  const rclcpp_lifecycle::State & /* previous_state */)
-{
-  RCLCPP_INFO(rclcpp::get_logger("ABBSystemPositionOnlyHardware"), "on_deactivate()");
+  RCLCPP_INFO(LOGGER, "CONNECTED!");
 
-  io_service_.stop();
-  thread_group_.join_all();
+  egm_manager_->read(motion_data_);
 
-  RCLCPP_INFO(rclcpp::get_logger("ABBSystemPositionOnlyHardware"), "System sucessfully stopped!");
+  RCLCPP_INFO(LOGGER, "System sucessfully started!");
 
   return CallbackReturn::SUCCESS;
 }
 
 return_type ABBSystemPositionOnlyHardware::read()
 {
-  RCLCPP_INFO(rclcpp::get_logger("ABBSystemPositionOnlyHardware"), "read()");
-
-  abb::egm::wrapper::Input temp_input;
-  abb::egm::wrapper::Joints temp_current_positions;
-  egm_interface_->read(&temp_input);
-  temp_current_positions.CopyFrom(temp_input.feedback().robot().joints().position());
-
-  for (size_t i = 0; i < hw_states_.size(); ++i) {
-    hw_states_[i] = temp_current_positions.values(i) / DEG_TO_RAD;
-    RCLCPP_INFO(
-      rclcpp::get_logger("ABBSystemPositionOnlyHardware"), "Got state %.5f for joint %ld!",
-      hw_states_[i], i);
+  RCLCPP_INFO(LOGGER, "read()");
+
+  if (!egm_manager_->read(motion_data_)) {
+    RCLCPP_INFO(LOGGER, "Failed to read EGM feedback - is EGM session active?");
+  } else {
+    for (auto & group : motion_data_.groups) {
+      for (auto & unit : group.units) {
+        for (auto & joint : unit.joints) {
+          RCLCPP_INFO(LOGGER, "Got state %.5f for %s!", joint.state.position, joint.name.c_str());
+        }
+      }
+    }
   }
-
   return return_type::OK;
 }
 
 return_type ABBSystemPositionOnlyHardware::write()
 {
-  RCLCPP_INFO(rclcpp::get_logger("ABBSystemPositionOnlyHardware"), "write()");
-
-  for (size_t i = 0; i < hw_commands_.size(); ++i) {
-    RCLCPP_INFO(
-      rclcpp::get_logger("ABBSystemPositionOnlyHardware"), "Got command %.5f for joint %ld!",
-      hw_commands_[i], i);
-    output_pos_.mutable_robot()->mutable_joints()->mutable_position()->set_values(
-      i, (hw_commands_[i] * 180.0 / 3.14159));
-  }
-  egm_interface_->write(output_pos_);
+  RCLCPP_INFO(LOGGER, "write()");
 
+  for (auto & group : motion_data_.groups) {
+    for (auto & unit : group.units) {
+      for (auto & joint : unit.joints) {
```

Comment:
Is this too nitpicky? There are some other places that could use the same change.

```suggestion
  for (const auto & group : motion_data_.groups) {
    for (const auto & unit : group.units) {
      for (const auto & joint : unit.joints) {
```